### PR TITLE
device-update convenience changes

### PR DESCRIPTION
**Pull request by srinivasaraogattupalli**
---
This project has adopted the [Microsoft Open Source Code of Conduct](https://opensource.microsoft.com/codeofconduct/). For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.

Thank you for contributing to the IoT extension!

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [ ] If introducing new functionality or modified behavior, are they backed by unit and integration tests?
- [ ] In the same context as above are command names and their parameter definitions accurate? Do help docs have sufficient content?
- [ ] Have **all** unit **and** integration tests passed locally? i.e. `pytest <project root> -vv`
- [ ] Have static checks passed using the .pylintrc and .flake8 rules? Look at the CI scripts for example usage.
- [ ] Have you made an entry in HISTORY.rst which concisely explains your feature or change?

### CODE REVIEW

File: azext_iot/tests/test_iot_ext_unit.py [hunk 1]
```diff
@@ -360,12 +360,32 @@ def generate_device_show(**kvp):
...
+        "deviceId": device_id,
+        "status": "disabled",
+    }
+    for k in kvp:
+        if payload.get(k):
+            payload[k] = kvp[k]
+    return payload
+
+
+def generate_device_show_custom(**kvp):
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_unit.py]: What is the difference with the previous function "generate_device_show"

File: azext_iot/tests/test_iot_ext_unit.py [hunk 2]
```diff
@@ -360,12 +360,32 @@ def generate_device_show(**kvp):
...
+    payload = {
+        "authentication": {
+            "symmetricKey": {"primaryKey": "123", "secondaryKey": "321"},
+            "x509Thumbprint": {"primaryThumbprint": "123", "secondaryThumbprint": "321"},
             "type": "sas",
+            "auth_method": "shared_private_key"
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_unit.py]: Why "type" and "auth_method" both exist

File: azext_iot/tests/test_iot_ext_unit.py [hunk 3]
```diff
@@ -431,6 +453,145 @@ def test_device_update(self, fixture_cmd, serviceclient, req):
...
+                    }
+                ),
+                generate_device_show_custom(
+                    authentication={
+                        "symmetricKey": {"primaryKey": "primarykeyUpdated", "secondaryKey": "primarykeyUpdated"},
+                        "type": "certificateAuthority",
+                        "auth_method": "x509_ca",
+                        "x509Thumbprint": {"primaryThumbprint": "123", "secondaryThumbprint": "321"},
+                    }
+                )
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_unit.py]: Why are we passing the same function as input

File: azext_iot/tests/test_iot_ext_int.py [hunk 4]
```diff
@@ -444,8 +444,53 @@ def test_hub_devices(self):
...
+                 .format(device_ids[0], LIVE_HUB, LIVE_RG, 'x509_thumbprint'),
+                 expect_failure=True)
+
+        self.cmd("iot hub device-identity update -d {} -n {} -g {} --auth-method {} --pk {}"
+                 .format(device_ids[0], LIVE_HUB, LIVE_RG, 'shared_private_key', '123'),
+                 expect_failure=True)
+
+        self.cmd(
+            '''iot hub device-identity update -d {} -n {} -g {} --primary-key=""
+                    --secondary-key=""'''.format(
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_int.py]: What is this test doing

**Review by anusapan — CHANGES_REQUESTED** [review]

**Comment by anusapan** [comment]
Are Flake and Linter passed?

File: azext_iot/tests/test_iot_ext_int.py [hunk 5]
```diff
@@ -444,29 +446,101 @@ def test_hub_devices(self):
...
+                self.check("authentication.symmetricKey.secondaryKey", SECONDARY_KEY),
             ],
         )
 
         # With connection string
         self.cmd(
-            '''iot hub device-identity update -d {} --login {} --set authentication.symmetricKey.primaryKey=""
-                 authentication.symmetricKey.secondaryKey=""'''.format(
-                edge_device_ids[1], self.connection_string
+            '''iot hub device-identity update -d {} --login {} --set authentication.symmetricKey.primaryKey={}
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_int.py]: We can change it to convenience argument

File: azext_iot/tests/test_iot_ext_int.py [hunk 6]
```diff
@@ -444,29 +446,101 @@ def test_hub_devices(self):
...
+                 authentication.symmetricKey.secondaryKey={}'''.format(
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_int.py]: same for this too

File: azext_iot/tests/test_iot_ext_int.py [hunk 7]
```diff
@@ -444,29 +446,101 @@ def test_hub_devices(self):
...
+        )
+
+        self.cmd(
+            "iot hub device-identity update -d {} -n {} -g {} --ptp {}"
+            .format(device_ids[0], LIVE_HUB, LIVE_RG, PRIMARY_THUMBPRINT),
+            checks=[
+                self.check("deviceId", device_ids[0]),
+                self.check("status", "enabled"),
+                self.check("authentication.x509Thumbprint.primaryThumbprint", PRIMARY_THUMBPRINT),
+                self.check("authentication.x509Thumbprint.secondaryThumbprint", PRIMARY_THUMBPRINT),
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_int.py]: How is that possible? checking Primary_Thumbprint to secondaryThuumbprint field. Did you get a pass to this test?

File: azext_iot/tests/test_iot_ext_int.py [hunk 8]
```diff
@@ -444,29 +446,101 @@ def test_hub_devices(self):
...
+
+        self.cmd("iot hub device-identity update -d {} -n {} -g {} --auth-method {} --pk {}"
+                 .format(device_ids[0], LIVE_HUB, LIVE_RG, 'shared_private_key', '123'),
+                 expect_failure=True)
+
+        # Swapping primary and secondary thumprint to check the update status
+        self.cmd(
+            "iot hub device-identity update -d {} -n {} -g {} --ptp {} --stp {}"
+            .format(device_ids[0], LIVE_HUB, LIVE_RG, SECONDARY_THUMBPRINT, PRIMARY_THUMBPRINT),
+            checks=[
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_int.py:502]: What we are trying to achieve for this test?

File: azext_iot/tests/test_iot_ext_int.py [hunk 9]
```diff
@@ -444,29 +446,101 @@ def test_hub_devices(self):
...
+            .format(device_ids[0], LIVE_HUB, LIVE_RG, PRIMARY_THUMBPRINT),
+            checks=[
+                self.check("deviceId", device_ids[0]),
+                self.check("status", "enabled"),
+                self.check("authentication.x509Thumbprint.primaryThumbprint", PRIMARY_THUMBPRINT),
+                self.check("authentication.x509Thumbprint.secondaryThumbprint", PRIMARY_THUMBPRINT),
+            ],
+        )
+
+        self.cmd("iot hub device-identity update -d {} -n {} -g {} --pk {}"
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_int.py:522]: make a comment here, what is this test doing...

File: azext_iot/tests/test_iot_ext_unit.py [hunk 10]
```diff
@@ -469,6 +496,142 @@ def test_device_update_error(self, serviceclient_generic_error, req):
...
+        [
+            generate_device_custom(generate_device_show(), False, "disabled"),
+            generate_device_custom(generate_device_show(), True, "enabled", "status_updated"),
+            generate_device_custom(
+                generate_device_show(
+                    authentication={
+                        "symmetricKey": {"primaryKey": "", "secondaryKey": ""},
+                        "type": "sas",
+                    }
+                ), True, "enabled", None, "shared_private_key", None, None, "primary_Updated", "secondary_Updated"
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_unit.py]: we can pass only those arguments which are required

File: azext_iot/tests/test_iot_ext_unit.py [hunk 11]
```diff
@@ -469,6 +496,142 @@ def test_device_update_error(self, serviceclient_generic_error, req):
...
+        if req["auth"] == "x509_ca":
+            assert instance['authentication']['type'] == "certificateAuthority"
+
+        if req["auth"] is None and (req["pk"] or req["sk"] or req["ptp"] or req["stp"]):
+            if instance['authentication']['type'] == 'sas':
+                assert instance['authentication']['symmetricKey']['primaryKey'] == req["pk"]
+                assert instance['authentication']['symmetricKey']['secondaryKey'] == req["sk"]
+            if instance['authentication']['type'] == 'selfSigned':
+                assert instance['authentication']['x509Thumbprint']['primaryThumbprint'] == req["ptp"]
+                assert instance['authentication']['x509Thumbprint']['secondaryThumbprint'] == req["stp"]
```
Comment (anusapan) [inline azext_iot/tests/test_iot_ext_unit.py]: What will happen we won't pass secondary thumbprint or req["stp"]=None, then it will fail

**Review by anusapan — CHANGES_REQUESTED** [review]

File: azext_iot/tests/test_iot_ext_unit.py (hunk 1 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_unit.py]: updated code as per our discussion and fixed

File: azext_iot/tests/test_iot_ext_unit.py (hunk 3 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_unit.py]: Fixed as per our discussion

File: azext_iot/tests/test_iot_ext_unit.py (hunk 2 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_unit.py]: Fixed as per discussion

File: azext_iot/tests/test_iot_ext_int.py (hunk 4 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_int.py]: Fixed as per discussion

File: azext_iot/tests/test_iot_ext_unit.py (hunk 2 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_unit.py]: Fixed as per discussion

**Comment by srinivasaraogattupalli** [comment]
> Are Flake and Linter passed?

yes all Flake and Linter got passed

File: azext_iot/tests/test_iot_ext_unit.py (hunk 10 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_unit.py]: updated as per suggested

File: azext_iot/tests/test_iot_ext_int.py (hunk 8 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_int.py:502]: we are swapping the keys and verifying them whether they updated or not

File: azext_iot/tests/test_iot_ext_int.py (hunk 9 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_int.py:522]: updated test case now

File: azext_iot/tests/test_iot_ext_int.py (hunk 7 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_int.py]: we are passing primary thumbprint to secondary and verifying the secondary value

File: azext_iot/tests/test_iot_ext_int.py (hunk 5 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_int.py]: Fixed

File: azext_iot/tests/test_iot_ext_int.py (hunk 6 above)
Comment (srinivasaraogattupalli) [reply azext_iot/tests/test_iot_ext_int.py]: Fixed

**Pull request closed without merging**